### PR TITLE
[Multi-task Learning] Refactor graphstorm.model for multi-task learning.

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#789 

*Description of changes:*
As multi-task learning trainer will invoke edge_mini_batch_predict, lp_mini_batch_predict and node_mini_batch_predict when conducting evaluation or testing, refactor the code to allow the functions to work with different decoders.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/model/edge_gnn.py [hunk 1]
```diff
@@ -311,6 +311,48 @@ def edge_mini_batch_predict(model, emb, loader, return_proba=True, return_label=
...
+                                                return_proba,
+                                                return_label)
+    model.train()
+    return preds, labels
+
+def run_edge_mini_batch_predict(decoder, emb, loader, device,
+                                return_proba=True, return_label=False):
+    """ Perform mini-batch prediction with the given decoder.
+
+    This function usually follows full-grain GNN embedding inference. After having
```
Comment (zhjwy9343) [inline python/graphstorm/model/edge_gnn.py:328]: "full-grain"?

Is it "full-graph"?

File: python/graphstorm/model/edge_gnn.py [hunk 2]
```diff
@@ -311,6 +311,48 @@ def edge_mini_batch_predict(model, emb, loader, return_proba=True, return_label=
...
+    the GNN embeddings, we need to perform mini-batch computation to make predictions
+    on the GNN embeddings.
+
+    Note: caller should call model.eval() before calling this function
```
Comment (zhjwy9343) [inline python/graphstorm/model/edge_gnn.py:332]: caller -> callers

File: python/graphstorm/model/edge_gnn.py [hunk 3]
```diff
@@ -311,6 +311,48 @@ def edge_mini_batch_predict(model, emb, loader, return_proba=True, return_label=
...
+                                                loader,
+                                                device,
+                                                return_proba,
+                                                return_label)
+    model.train()
+    return preds, labels
+
+def run_edge_mini_batch_predict(decoder, emb, loader, device,
+                                return_proba=True, return_label=False):
+    """ Perform mini-batch prediction with the given decoder.
```
Comment (zhjwy9343) [inline python/graphstorm/model/edge_gnn.py:326]: "mini-batch prediction" -> "mini-batch edge prediction"

File: python/graphstorm/model/lp_gnn.py [hunk 4]
```diff
@@ -154,6 +154,37 @@ def lp_mini_batch_predict(model, emb, loader, device):
...
+                                     device)
+
+def run_lp_mini_batch_predict(decoder, emb, loader, device):
+    """ Perform mini-batch link prediction with the given decoder.
+
+        This function follows full-graph GNN embedding inference.
+        After having the GNN embeddings, we need to perform mini-batch
+        computation to make predictions on the GNN embeddings.
+
+        Note: caller should call model.eval() before calling this function
```
Comment (zhjwy9343) [inline python/graphstorm/model/lp_gnn.py:169]: Should we add model.eval() between line #156 and #157, and add model.train() after calling this method in the line #157?

File: python/graphstorm/model/node_gnn.py [hunk 5]
```diff
@@ -311,6 +311,48 @@ def node_mini_batch_predict(model, emb, loader, return_proba=True, return_label=
...
+                                    loader,
+                                    device,
+                                    return_proba,
+                                    return_label)
+    model.train()
+    return preds, labels
+
+def run_node_mini_batch_predict(decoder, emb, loader, device,
+                                return_proba=True, return_label=False):
+    """ Perform mini-batch prediction with the given decoder.
```
Comment (zhjwy9343) [inline python/graphstorm/model/node_gnn.py:328]: "mini-batch prediction" -> "mini-batch node prediction"

File: python/graphstorm/model/lp_gnn.py (hunk 4 above)
Comment (zhjwy9343) [inline python/graphstorm/model/lp_gnn.py:169]: caller -> callers

File: python/graphstorm/model/node_gnn.py [hunk 6]
```diff
@@ -311,6 +311,48 @@ def node_mini_batch_predict(model, emb, loader, return_proba=True, return_label=
...
+
+        Note: caller should call model.eval() before calling this function
```
Comment (zhjwy9343) [inline python/graphstorm/model/node_gnn.py:330]: caller -> callers

File: python/graphstorm/model/lp_gnn.py (hunk 4 above)
Comment (classicsong) [reply python/graphstorm/model/lp_gnn.py:169]: In lp_mini_batch_predict, we only use the model decoder, so we do not know whether the caller will do something else for evaluation after calling lp_mini_batch_predict.

Let me add some notes in the function.

Comment (zhjwy9343) [reply python/graphstorm/model/lp_gnn.py]: I C.

**Review by zhjwy9343 — APPROVED** [review]
LGTM

**Pull request merged**